internal/sync/lifecycle.go

```diff
@@ -117,39 +101,45 @@ func Once() LifecycleOnce {
 // If Start is called multiple times it will return the error
 // from the first time it was called.
 func (l *lifecycleOnce) Start(f func() error) error {
-	if l.starting.Swap(true) {
-		<-l.startCh
+	if l.state.CAS(int32(Idle), int32(Starting)) {
+		if f != nil {
+			l.err = f()
+		}
+
+		// skip forward to error state
+		if l.err != nil {
+			l.state.Store(int32(Errored))
+			close(l.stopCh)
+		} else {
+			l.state.Store(int32(Running))
+		}
+		close(l.startCh)
 		return l.err
 	}
-	if f != nil {
-		l.err = f()
-	}
-	// skip forward to error state
-	if l.err != nil {
-		l.errored.Store(true)
-		l.stopped.Store(true)
-		l.stopping.Store(true)
-		close(l.stopCh)
-	}
-	l.started.Store(true)
-	close(l.startCh)
 
+	<-l.startCh
```

Comment:
What about moving this condition variable wait in the else clause of the CAS condition? You only need to wait on the starting condition variable if you do not own the starting process. Of course the current code works, but for cleanliness you know.